src/Processors/LimitTransform.cpp

```diff
@@ -257,89 +256,161 @@ void LimitTransform::splitChunk(PortsData & data)
+            start = offset + num_rows - rows_read;
 
-    if (offset + num_rows > rows_read)
-        start = offset + num_rows - rows_read;
+        /// ------------[....(...).]
+        /// <----------------------> rows_read
+        ///             <----------> num_rows
+        /// <---------------> offset
+        ///                  <---> limit
+        ///                  <---> length
+        ///             <---> start
 
-    /// ------------[....(...).]
-    /// <----------------------> rows_read
-    ///             <----------> num_rows
-    /// <---------------> offset
-    ///                  <---> limit
-    ///                  <---> length
-    ///             <---> start
+        /// Or:
 
-    /// Or:
+        /// -----------------(------[....)....]
+        /// <---------------------------------> rows_read
+        ///                         <---------> num_rows
+        /// <---------------> offset
+        ///                  <-----------> limit
+        ///                         <----> length
+        ///                         0 = start
 
-    /// -----------------(------[....)....]
-    /// <---------------------------------> rows_read
-    ///                         <---------> num_rows
-    /// <---------------> offset
-    ///                  <-----------> limit
-    ///                         <----> length
-    ///                         0 = start
+        length = num_rows - start;
 
-    UInt64 length = num_rows - start;
+        if (!limit_is_unreachable && offset + limit < rows_read)
+        {
+            if (offset + limit < rows_read - num_rows)
+                length = 0;
+            else
+                length = offset + limit - (rows_read - num_rows) - start;
+        }
 
-    if (!limit_is_unreachable && offset + limit < rows_read)
-    {
-        if (offset + limit < rows_read - num_rows)
-            length = 0;
-        else
-            length = offset + limit - (rows_read - num_rows) - start;
-    }
+        /// check if other rows in current block equals to last one in limit
+        if (with_ties && length)
+        {
+            UInt64 current_row_num = start + length;
+            previous_row_chunk = makeChunkWithPreviousRow(data.current_chunk, current_row_num - 1);
+
+            for (; current_row_num < num_rows; ++current_row_num)
+            {
+                if (!sortColumnsEqualAt(current_chunk_sort_columns, current_row_num))
+                {
+                    previous_row_chunk = {};
+                    break;
+                }
+            }
 
-    /// check if other rows in current block equals to last one in limit
-    if (with_ties && length)
+            length = current_row_num - start;
+        }
+    }
+    else
     {
-        UInt64 current_row_num = start + length;
-        previous_row_chunk = makeChunkWithPreviousRow(data.current_chunk, current_row_num - 1);
+        limit_is_unreachable = (limit > std::numeric_limits<UInt64>::max() - offset);
```

Comment:
This line is exactly the same as line 265, we can move it before if/else statement I guess